README.txt
Introduction
============

**Makes Plone File Uploads easy**

Multifileupload for Plone using uploadify_

.. _uploadify: http://www.uploadify.com

Usage
*****

After insall, go to http://your-plone-site/@@upload


Integration in Plone
********************

.. sidebar:: Note

    collective.uploadify contains **no** GenericSetup Profile, thus, it won't
    appear in the quickinstaller tool.

If you want to smoothly integrate the upload funtionality to your site,
consider to add the following lines to your policy product in the
profiles/default/actions.xml::

    <?xml version="1.0"?>
    <object name="portal_actions" meta_type="Plone Actions Tool"
       xmlns:i18n="http://xml.zope.org/namespaces/i18n">

     <!-- *** OBJECT *** -->
     <object name="object" meta_type="CMF Action Category">
      <property name="title"></property>

      <!-- MULTI UPLOAD -->
      <object name="upload" meta_type="CMF Action" i18n:domain="collective.uploadify">
       <property name="title" i18n:translate="">Upload</property>
       <property name="description" i18n:translate="">Batch upload files.</property>
       <property name="url_expr">string:${folder_url}/@@upload</property>
       <property name="icon_expr"></property>
       <property name="available_expr"></property>
       <property name="permissions">
        <element value="Add portal content"/>
       </property>
       <property name="visible">True</property>
      </object>
     </object>
    </object>

or simply go to the Zope Management Interface -> portal_actions -> object and
add a new CMF Action Category from the dropdown and configure it with the
following lines::

    URL (Expression): string:${folder_url}/@@upload


Configuration
*************

The following settings can be done in the site_properties.
(please use **string** properties only!):

  - ul_auto_upload -- true/false (default: false)

    *Set to true if you would like the files to be uploaded when they are
    selected.*

  - ul_allow_multi -- true/false (default: true)

    *Set to true if you want to allow multiple file uploads.*

  - ul_sim_upload_limit -- number 1-n (default: 4)

    *A limit to the number of simultaneous uploads you would like to allow.*

  - ul_queue_size_limit -- number 1-n (default: 999)

    *A limit to the number of files you can select to upload in one go.*

  - ul_size_limit -- size in bytes (default: empty)

    *A number representing the limit in bytes for each upload.*

  - ul_file_description -- text (default: empty)

    *The text that will appear in the file type drop down at the bottom of the
    browse dialog box.*

  - ul_file_extensions -- list of extensions (default: \*.\*;)

    *A list of file extensions you would like to allow for upload.  Format like
    \*.ext1;\*.ext2;\*.ext3. FileDesc is required when using this option.*

  - ul_button_text -- text (default: BROWSE)

    *The text you would like to appear on the default button.*

  - ul_button_image -- path to image (default: empty)

    *The path to the image you will be using for the browse button.
    NOTE: If you are using a **different sized button image** you have to set
    ul_height and ul_width otherwise your ul_button_image will be stretched to
    the defaults (110x30)*

  - ul_hide_button -- true/false (default: false)

    *Set to true if you want to hide the button image.*

  - ul_script_access -- always/sameDomain (default: sameDomain)

    *The access mode for scripts in the flash file.  If you are testing locally, set to `always`.*

  - ul_width -- number (default: 110)

    *The ul_width value which should be set when using a different sized
    ul_button_image*

  - ul_height -- number (default: 30)

    *The ul_height value which should be set when using a different sized
    ul_button_image*


**NEW in version 1.0rc2:**

  - ul_scale_image_size -- x,y

    *These two values define the max x,y size in pixels of the image. Scales
    an image down to at most ul_scale_image_size size preserving aspect ratio.
    Example: 800,600 to set a maximum size of 800x600 pixels*



Adding a custom File Mutator Utility
************************************

If you want to so some special handling for uploaded files *before* they get
created in the portal, you can simply register a new utility providing the
IFileMutator Interface.

Your utility will be called with **file_name, file_data, content_type** just
before the content will be created in the portal.

.. sidebar:: Parameters

    **file_name**::

        type: str

        example: my-image.jpg

    **file_data**::

        type: <ZPublisher.HTTPRequest.FileUpload instance at -...>

        can be used just like a file.

    **content_type**::

        type: str

        example: 'image/jpeg'


Example
-------

A simple utility which adds a "photo-" prefix to image filenames


configure.zcml::

    <!-- An Utility to give images an "photo-" prefix -->
    <utility component=".utility.prefix_image_filename"
             name="prefix-image-filename"/>


utility.py::

    from zope import interface

    def prefix_image_filename(file_name, file_data, content_type):
        """ Prefix all images with 'photo-<filename>'
        """
        # we only handle image files
        if not content_type.startswith("image"):
            return (file_name, file_data, content_type)

        if not file_name.startswith("photo"):
            file_name = "-".join(["photo", file_name])
        return (file_name, file_data, content_type)


    interface.directlyProvides(prefix_image_filename,
                               IFileMutator)


.. note::

    Your utility has to return a tuple of::

        (file_name, file_data, content_type)


Customization for specific BrowserLayer
***************************************

If you have your own skin installed which is based on it's own BrowserLayer
(by default IThemeSpecific) and want to customize the look and feel of
collective.uploadify's you could override the upload view and/or the upload
initialize callback view

 - Customize view template::

     <configure
         xmlns="http://namespaces.zope.org/zope"
         xmlns:browser="http://namespaces.zope.org/browser">

         ...

         <!-- Customize collective.uploadify upload template -->
         <browser:page
             for="collective.uploadify.browser.interfaces.IUploadingCapable"
             name="upload"
             template="templates/upload.pt"
             permission="cmf.AddPortalContent"
             layer=".interfaces.IThemeSpecific"
             />

     </configure>

 - Customize initialize template::

    from zope.i18n import translate
    from zope.component import getMultiAdapter
    from collective.uploadify.browser.upload import UploadInitalize
    from my.product import MyMessageFactory as _


    class MyCustomUploadInitalize(UploadInitalize):
        """ Initialize uploadify js
        """

        def upload_settings(self):

            portal_state = getMultiAdapter(
                (self.context, self.request), name="plone_portal_state")

            settings = super(MyCustomUploadInitalize, self).upload_settings()
            settings.update(dict(
                ul_height = 26,
                ul_width = 105,
                ul_button_text = translate(_('Choose images'), target_language= \
                                           self.request.get('LANGUAGE', 'de')),
                ul_button_image = portal_state.navigation_root_url() + \
                    '/button_upload.png',
            ))

            return settings

   Don't forget to register the new upload initialize view for your
   themespecific browserlayer by using::

     <configure
          xmlns="http://namespaces.zope.org/zope"
          xmlns:browser="http://namespaces.zope.org/browser">

        ...

         <browser:page
             for="*"
             name="upload_initialize"
             class=".uploadify.MyCustomUploadInitalize"
             permission="cmf.AddPortalContent"
             layer=".interfaces.IThemeSpecific"
             />

     </configure>


docs/HISTORY.txt
Changelog
*********

1.3 (unreleased)
----------------

* Add uninstallation profile.
  [thet]


1.2 (2013-05-11)
----------------

* make compatible with dexterity (plone.app.contenttypes)
  [pbauer]

* use IUserPreferredURLNormalizer to set id's
  [pbauer]


1.1 - 2012-10-11
----------------

* moved to GitHub
  [ramonski]

* Plone 4.3 compatibility
  [alert]

* added italian translation.
  [keul]

* cleanup egg format, removing bad zopeskel elements
  [keul]

* added default genericsetup profile which adds folder tab for on each folder.
  [garbas]

1.0 - 2011-05-05
----------------

* fixed the loss of File Extensions for Plone > 3.3.5 with IDNormalizer.
  Changed to use the IFileNameNormalizer Utility [jakke, ramonski]
* added dutch translation [spereverde]

1.0rc5 - 2011-04-04
-------------------

* Support for different contenttypes [thor27]
* added basque translation [erral]
* added spain translation [macagua]

1.0rc4 - 2011-01-31
-------------------

* Added German translations [fRiSi]

* Added Chinese (Taiwan)  translations [marr]

* Updated documentation to show Upload action on default
  pages too [fRiSi]

1.0rc3 - 2010-09-18
-------------------

* Fire Archetypes ObjectInitializedEvent for uploaded content.
  [hannosch]

* Corrected ``i18n:domain`` in documented action and include its messages into
  the pot file.
  [hannosch]

* Added a ``z3c.autoinclude`` entry point to avoid the need for a ZCML slug in
  Plone 3.3 and later.
  [hannosch]

* changed permissions to cmf.AddPortalContent
  this fixes http://plone.org/products/collective.uploadify/issues/13
  [ramonski]

* mimetypes.guess_type(file_name) returns now an empty string if the type
  could not be guessed. This fixes an issue with the scale_image_size utility
  [ramonski]

* documentation updated
  [ramonski]

1.0rc2 - 2010-08-03
-------------------

* Interface "IFileMutator" introduced -- utilities which provide this
  interface can do some file mutations on the uploaded file *before* it get
  created as an content type
  [ramonski]

* utility scale_image_size registered -- this utility operates on images an
  can scale down the image size using PIL preserving the aspect ration from
  the input file
  [ramonski]

* id chooser gets now called *after* the thread lock. This should fix #8
  [ramonski]

* added adapter for cStringIO.OutputType -> IBlobbable to be compatible with
  plone.app.blob. This adapter gets only registered when plone.app.blob is
  istalled.
  [ramonski]

* use folder_url and remove is_folderish() and wf-state condition
  to make upload action show up on private folders or
  folders having a default page too
  (fixes http://plone.org/products/collective.uploadify/issues/12)
  [fRiSi]

* i18n added + danish translation
  [stonor]

1.0rc1 - 2010-04-25
-------------------

* renamed resource swfobject.js to uploadify.swfobject.js.
  This fixes the name clash with quintagroup.portlet.cumulus
  [ramonski]

* Added some basic tests to the package
  [ramonski]

* Added queue size to the settings in UploadInitialize flash response to
  give users the possibility to customize their upload queue size.
  [zupo]

* Documentation updates. Added new section about customize uploadify's upload
  view and upload initialize view templates.
  [saily]

* Added height and width to the settings in UploadInitialize flash response to
  give users the possibility to customize their uploadbuttons.
  [saily]

* #9 fire ObjectModifiedEvent after upload of each item.
  (fixes http://plone.org/products/collective.uploadify/issues/9)
  [saily]

0.10 - 2010-03-16
-----------------

* #5 play well with plone4's sunburst theme
  (fixes http://plone.org/products/collective.uploadify/issues/5)
  [fRiSi]

* #7 reindex objects after upload so uploaded image files get
  recognized as images (fixes http://plone.org/products/collective.uploadify/issues/7)
  [fRiSi]

0.9 - 2009-10-29
----------------

* updated to Uploadify v2.1.0
  [ramonski]

* removed cgi parser for QUERY_STRING.
  Now we receive the data via POST
  [ramonski]

* updated js, css and upload.pt
  [ramonski]

0.8 - 2009-10-27
----------------

* added cgi parser for QUERY_STRING
  [seletz, ramonski]

* added custom string encoder/decoder
  [seletz, ramonski]

* fixed name chooser for german umlauts
  [ramonski]

0.7 - 2009-07-23
----------------

* re-fixed name chooser ...gnaarf
  [ramonski]

* removed silly logging
  [ramonski]

0.6 - 2009-07-23
----------------

* changed the paths to the resources in javascript from absolute to relative.
  This fixes some strange behaviour resulting in Uploadify - IOErrors
  [ramonski]

* fixed #2 (AttributeError: aq_parent) -- when doing uploads in the portal root
  [ramonski]

* The upload folder can now be in private state as well.
  [ramonski]

0.5 - 2009-06-25
----------------

* use now the jquery.js shipped with Plone.
  This satisfies the comment of Maurits van Rees in
  http://old.plone.org/products/ploneflashupload/issues/1
  [ramonski]

* fixed Bug with M$ IE Browsers described in
  http://www.uploadify.com/forum/viewtopic.php?f=4&t=266
  [ramonski]

* uses 'jq' variable instead of '$' in templates and javascript
  [ramonski]

0.4 - 2009-06-23
----------------

* added cancel button
  [ramonski]

* uploadify settings can be set in site_properties
  [ramonski]

* added a name chooser to avoid id clashes when uploading files
  [ramonski]

0.3 - 2009-06-22
----------------

* fixed bug for M$ IE Browsers that the 'Browse' button not appear
  [ramonski]

* changed permission for upload view to cmf.ModifyPortalContent
  [ramonski]

* added onAllComplete handler which reloads the location and displays the
  uploaded files immediatly in the folder_listing macro
  [ramonski]

* registered a browser view for the uploadify javascript initialization
  [ramonski]

* registered a browser view for the uploadify upload action
  [ramonski]

* removed unused code
  [ramonski]

0.2.2 - 2009-06-04
------------------


collective/uploadify/browser/atct.py
# -*- coding: utf-8 -*-
#
# File: __init__.py
#
# Copyright (c) InQuant GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = 'Ramon Bartl <ramon.bartl@inquant.de>'
__docformat__ = 'plaintext'

import transaction
from thread import allocate_lock

from zope import interface
from zope import component

from zope.event import notify
from zope.lifecycleevent import ObjectModifiedEvent
from zope.filerepresentation.interfaces import IFileFactory

# Since Plone4.3 INameChooser moved to zope.container
try:
    from zope.container.interfaces import INameChooser
except:
    from zope.app.container.interfaces import INameChooser

from plone.i18n.normalizer.interfaces import IFileNameNormalizer

from Products.Archetypes.event import ObjectInitializedEvent
from Products.CMFPlone import utils as ploneutils
from Products.CMFCore import utils as cmfutils

from interfaces import IUploadingCapable

upload_lock = allocate_lock()


class UploadingCapableFileFactory(object):
    interface.implements(IFileFactory)
    component.adapts(IUploadingCapable)

    def __init__(self, context):
        self.context = context

    def __call__(self, name, content_type, data):
        ctr = cmfutils.getToolByName(self.context, 'content_type_registry')
        type_ = ctr.findTypeName(name.lower(), '', '') or 'File'

        # XXX: quick fix for german umlauts
        name = name.decode("utf8")

        normalizer = component.getUtility(IFileNameNormalizer)
        chooser = INameChooser(self.context)

        # otherwise I get ZPublisher.Conflict ConflictErrors
        # when uploading multiple files
        upload_lock.acquire()

        # this should fix #8
        newid = chooser.chooseName(normalizer.normalize(name), self.context.aq_parent)
        try:
            transaction.begin()
            obj = ploneutils._createObjectByType(type_, self.context, newid)
            mutator = obj.getPrimaryField().getMutator(obj)
            mutator(data, content_type=content_type)
            obj.setTitle(name)
            obj.reindexObject()

            notify(ObjectInitializedEvent(obj))
            notify(ObjectModifiedEvent(obj))

            transaction.commit()
        finally:
            upload_lock.release()
        return obj

# vim: set ft=python ts=4 sw=4 expandtab :


collective/uploadify/browser/dexterity.py
# -*- coding: utf-8 -*-
#
# File: __init__.py

__author__ = 'Ramon Bartl <ramon.bartl@nexiles.de>'
__docformat__ = 'plaintext'

from Products.Archetypes.event import ObjectInitializedEvent
from Products.CMFCore import utils as cmfutils
from Products.CMFCore.utils import getToolByName
from Products.CMFPlone import utils as ploneutils
from plone.namedfile.file import NamedFile
from plone.namedfile.file import NamedImage
from plone.namedfile.interfaces import INamedFileField
from plone.namedfile.interfaces import INamedImageField
from plone.rfc822.interfaces import IPrimaryField
from thread import allocate_lock
from zope import component
from zope import interface
from zope.event import notify
from zope.filerepresentation.interfaces import IFileFactory
from zope.lifecycleevent import ObjectModifiedEvent
from zope.schema import getFieldsInOrder

from interfaces import IUploadingCapable

import transaction
import logging

try:
    from plone.namedfile.interfaces import INamedBlobFileField, INamedBlobImageField
    from plone.namedfile.file import NamedBlobFile, NamedBlobImage
    HAVE_BLOBS = True
except:
    HAVE_BLOBS = False

logger = logging.getLogger("collective.uploadify")
upload_lock = allocate_lock()


class UploadingCapableFileFactory(object):
    interface.implements(IFileFactory)
    component.adapts(IUploadingCapable)

    def __init__(self, context):
        self.context = context

    def __call__(self, name, content_type, data, obj_id):
        ctr = cmfutils.getToolByName(self.context, 'content_type_registry')
        type_ = ctr.findTypeName(name.lower(), '', '') or 'File'

        # otherwise I get ZPublisher.Conflict ConflictErrors
        # when uploading multiple files
        upload_lock.acquire()

        try:
            transaction.begin()
            obj = ploneutils._createObjectByType(type_, self.context, obj_id)

            ttool = getToolByName(self.context, 'portal_types')
            ctype = ttool[obj.portal_type]
            schema = ctype.lookupSchema()
            fields = getFieldsInOrder(schema)
            file_fields = [field for safe_name, field in fields
                           if INamedFileField.providedBy(field)
                           or INamedImageField.providedBy(field)]
            if len(file_fields) == 0:
                logger.info("An error happens : the dexterity content type %s "
                            "has no file field, rawdata can't be created",
                            obj.absolute_url())
            for file_field in file_fields:
                if IPrimaryField.providedBy(file_field):
                    break
            else:
                # Primary field can't be set ttw,
                # then, we take the first one
                file_field = file_fields[0]

            # TODO: use adapters
            if HAVE_BLOBS and INamedBlobImageField.providedBy(file_field):
                value = NamedBlobImage(data=data.read(), contentType=content_type,
                                       filename=unicode(obj_id, 'utf-8'))
            elif HAVE_BLOBS and INamedBlobFileField.providedBy(file_field):
                value = NamedBlobFile(data=data.read(), contentType=content_type,
                                      filename=unicode(obj_id, 'utf-8'))
            elif INamedImageField.providedBy(file_field):
                value = NamedImage(data=data.read(), contentType=content_type,
                                   filename=unicode(obj_id, 'utf-8'))
            elif INamedFileField.providedBy(file_field):
                value = NamedFile(data=data.read(), contentType=content_type,
                                  filename=unicode(obj_id, 'utf-8'))

            file_field.set(obj, value)
            obj.title = name
            obj.reindexObject()

            notify(ObjectInitializedEvent(obj))
            notify(ObjectModifiedEvent(obj))

            transaction.commit()
        finally:
            upload_lock.release()
        return obj


collective/uploadify/browser/upload.py
# -*- coding: utf-8 -*-
#
# File: upload.py
#
# Copyright (c) InQuant GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = 'Ramon Bartl <ramon.bartl@inquant.de>'
__docformat__ = 'plaintext'

import logging
import mimetypes

from Acquisition import aq_inner

from zope.filerepresentation.interfaces import IFileFactory

from Products.CMFCore.utils import getToolByName
from Products.Five.browser import BrowserView
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile

logger = logging.getLogger("collective.uploadify")


def encode(s):
    """ encode string
    """

    return "d".join(map(str, map(ord, s)))


def decode(s):
    """ decode string
    """

    return "".join(map(chr, map(int, s.split("d"))))


# NOTE, THIS IS NOT A PYTHON DICT:
# NEVER ADD A COMMA (,) AT THE END OF THE LAST KEY/VALUE PAIR, THIS BREAKS ALL
# M$ INTERNET EXPLORER
UPLOAD_JS = """
    function all_complete(event, data) {
        //alert(data.filesUploaded + " Files Uploaded!");
        //alert(data.errors + " Errors");
        //alert(data.speed + " Avg. Speed");
        location.reload();
    };
    jq(document).ready(function() {
        jq('#uploader').uploadify({
            'uploader'      : '++resource++uploader.swf',
            'script'        : '@@upload_file',
            'cancelImg'     : '++resource++cancel.png',
            'height'        : '%(ul_height)s',
            'width'         : '%(ul_width)s',
            'folder'        : '%(physical_path)s',
            'scriptData'    : {'cookie': '%(cookie)s'},
            'onAllComplete' : all_complete,
            'auto'          : %(ul_auto_upload)s,
            'multi'         : %(ul_allow_multi)s,
            'simUploadLimit': '%(ul_sim_upload_limit)s',
            'sizeLimit'     : '%(ul_size_limit)s',
            'fileDesc'      : '%(ul_file_description)s',
            'fileExt'       : '%(ul_file_extensions)s',
            'buttonText'    : '%(ul_button_text)s',
            'buttonImg'     : '%(ul_button_image)s',
            'scriptAccess'  : '%(ul_script_access)s',
            'hideButton'    : %(ul_hide_button)s
        });
    });
"""


class UploadView(BrowserView):
    """ The Upload View
    """

    template = ViewPageTemplateFile("upload.pt")

    def __call__(self):
        return self.template()


class UploadFile(BrowserView):
    """ Upload a file
    """

    def __init__(self, context, request):
        self.context = context
        self.request = request

        cookie = self.request.form.get("cookie")
        if cookie:
            self.request.cookies["__ac"] = decode(cookie)

    def __call__(self):

        file_name = self.request.form.get("Filename", "")
        file_data = self.request.form.get("Filedata", None)
        content_type = mimetypes.guess_type(file_name)[0]

        if file_data:
            factory = IFileFactory(self.context)
            logger.info("uploading file: filename=%s, content_type=%s" % \
                    (file_name, content_type))
            f = factory(file_name, content_type, file_data)
            logger.info("file url: %s" % f.absolute_url())
            return f.absolute_url()


class UploadInitalize(BrowserView):
    """ Initialize uploadify js
    """

    def __init__(self, context, request):
        super(UploadInitalize, self).__init__(context, request)
        self.context = aq_inner(context)

    def upload_settings(self):
        sp = getToolByName(self.context, "portal_properties").site_properties
        portal_url = getToolByName(self.context, 'portal_url')()

        settings = dict(
            cookie              = self.request.cookies.get('__ac', ''),
            portal_url          = portal_url,
            context_url         = self.context.absolute_url(),
            physical_path       = "/".join(self.context.getPhysicalPath()),
            ul_height           = sp.getProperty('ul_height', '30'),
            ul_width            = sp.getProperty('ul_width', '110'),
            ul_auto_upload      = sp.getProperty('ul_auto_upload', 'false'),
            ul_allow_multi      = sp.getProperty('ul_allow_multi', 'true'),
            ul_sim_upload_limit = sp.getProperty('ul_sim_upload_limit', 4),
            ul_size_limit       = sp.getProperty('ul_size_limit', ''),
            ul_file_description = sp.getProperty('ul_file_description', ''),
            ul_file_extensions  = sp.getProperty('ul_file_extensions', '*.*;'),
            ul_button_text      = sp.getProperty('ul_button_text', 'BROWSE'),
            ul_button_image     = sp.getProperty('ul_button_image', ''),
            ul_hide_button      = sp.getProperty('ul_hide_button', 'false'),
            ul_script_access    = sp.getProperty('ul_script_access', 'sameDomain'),
        )

        # This encoding is needed when the cookie value contains spaces!
        settings["cookie"] = encode(settings["cookie"])

        return settings

    def __call__(self):
        settings = self.upload_settings()
        return UPLOAD_JS % settings

# vim: set ft=python ts=4 sw=4 expandtab :
